File: leetcode/1312. Minimum Insertion Steps to Make a String Palindrome.rs
use std::cmp;

impl Solution {
    pub fn min_insertions(s: String) -> i32 {
        let v1: Vec<char> = s.chars().collect();
        let v2: Vec<char> = s.chars().rev().collect();
        
        let mut dp = vec![vec![0; s.len()+1]; s.len()+1];
        
        for i in (1..s.len()+1) {
            dp[i][0] = i;
            dp[0][i] = i;
        }
        
        for i in (1..v1.len()+1) {
            for j in (1..v2.len()+1) {
                if v1[i-1] == v2[j-1] {
                    dp[i][j] = dp[i-1][j-1]
                } else {
                    dp[i][j] = cmp::min(dp[i-1][j], dp[i][j-1]) + 1
                }
            }
        }
        
        (dp[s.len()][s.len()] / 2) as i32
    }
}

File: leetcode/1312. Minimum Insertion Steps to Make a String Palindrome.py
class Solution:
    def minInsertions(self, s: str) -> int:
        s1 = s
        s2 = s[::-1]
        
        dp = [[0 for _ in range(len(s)+1)] for _ in range(len(s)+1)]
        
        for i in range(1, len(s)+1):
            dp[0][i] = i
            dp[i][0] = i
            
        for i in range(len(s1)):
            for j in range(len(s2)):
                if s1[i] == s2[j]:
                    dp[i+1][j+1] = dp[i][j]
                else:
                    dp[i+1][j+1] = min(dp[i+1][j], dp[i][j+1]) + 1
        
        return dp[-1][-1] // 2
        
